src/sys/mod.rs

```diff
@@ -186,18 +192,46 @@ impl std::fmt::Debug for Poll {
 }
 
 impl Poll {
-    pub(crate) fn new() -> crate::Result<Poll> {
+    pub(crate) fn new(high_precision: bool) -> crate::Result<Poll> {
         Ok(Poll {
-            poller: Poller::new()?,
+            poller: Poller::new(high_precision)?,
             tokens: VecMap::new(),
+            timers: Arc::new(Mutex::new(TimerWheel::new())),
         })
     }
 
     pub(crate) fn poll(
         &mut self,
-        timeout: Option<std::time::Duration>,
+        mut timeout: Option<std::time::Duration>,
     ) -> crate::Result<Vec<PollEvent>> {
-        self.poller.poll(timeout)
+        let now = std::time::Instant::now();
+        // adjust the timeout for the timers
+        if let Some(next_timeout) = self.timers.lock().unwrap().next_deadline() {
```

Comment:
I think this is also another potentially blocking operation in the loop - see my suggestion elsewhere to use `try_unlock()` if you have a reason to think this should never block in actual use.

Reply:
The only other places where this lock is acquired are in the register/reregister/unregister methods of `Timer`, and this lock is acquired there for a short while (just the time to update the timer info). Given the loop and its handles are not threadsafe, I guess this can actually be replaced by a `Rc<RefCell<_>>`. That would make `Timer` non threadsafe as well, but tbh I don't think that is a problem, there is no point in sending a `Timer` between threads before inserting it in an event loop.